File: pkg/sortspec/sortspec.go
// Package sortspec parses, validates, and serializes multi-level sort
// specifications for the admin list endpoints (e.g. "side:asc,name:asc"). The
// grammar is deliberately self-contained (no DB, no HTTP): the binder validates
// a spec at bind time (see the sortspec validator in pkg/binder), the parties
// service turns a spec into ORDER BY clauses, and the React frontend mirrors the
// same grammar in app/libraries/sortSpec.ts. Which fields are sortable is a
// per-entity concern kept in fields.go.
package sortspec

import (
	"fmt"
	"strings"

	"github.com/pkg/errors"
)

// Direction is "asc" or "desc".
type Direction string

const (
	DirAsc  Direction = "asc"
	DirDesc Direction = "desc"
)

// MaxLevels is the hard cap on how many levels a spec may contain. It bounds the
// ORDER BY a single request can build; the admin lists have only a handful of
// sortable fields, so the cap is never reached in practice and exists to reject
// pathological input. Keep in sync with MAX_SORT_LEVELS in app/libraries/sortSpec.ts.
const MaxLevels = 8

// SortLevel is one field+direction pair in a spec. Field is one of the tokens in
// fields.go; the caller (Parse) has already checked it against the entity whitelist.
type SortLevel struct {
	Field     string
	Direction Direction
}

// Parse reads a serialized spec string (e.g. "side:asc,name:desc") into a slice
// of SortLevel, validating each field with isValidField. It rejects unknown
// fields, bad directions, duplicates, empty pairs, stray whitespace, and specs
// longer than MaxLevels. An empty string is an error (callers that treat "no
// sort" as valid check for "" before calling, e.g. ResolveParties).
//
// Mirrors parseSortSpec in app/libraries/sortSpec.ts (which returns null where
// this returns an error). Keep the two grammars in sync.
func Parse(s string, isValidField func(string) bool) ([]SortLevel, error) {
	if s == "" {
		return nil, errors.New("sort spec is empty")
	}
	// Whitespace is not allowed anywhere: this is a machine-readable URL param,
	// not human prose. Rejecting early keeps the grammar strict.
	if strings.ContainsAny(s, " \t\n\r") {
		return nil, errors.New("sort spec must not contain whitespace")
	}

	parts := strings.Split(s, ",")
	if len(parts) > MaxLevels {
		return nil, errors.Errorf("sort spec has %d levels, max is %d", len(parts), MaxLevels)
	}

	seen := make(map[string]struct{}, len(parts))
	levels := make([]SortLevel, 0, len(parts))

	for _, part := range parts {
		if part == "" {
			return nil, errors.New("sort spec contains an empty pair")
		}

		// SplitN with n=2 plus the length check rejects trailing junk like
		// "name:asc:extra" instead of silently dropping it.
		kv := strings.SplitN(part, ":", 2)
		if len(kv) != 2 {
			return nil, errors.Errorf("sort level %q missing direction", part)
		}

		field, dir := kv[0], kv[1]
		if !isValidField(field) {
			return nil, errors.Errorf("unknown sort field %q", field)
		}
		if dir != string(DirAsc) && dir != string(DirDesc) {
			return nil, errors.Errorf("invalid direction %q (want asc or desc)", dir)
		}
		if _, dup := seen[field]; dup {
			return nil, errors.Errorf("duplicate sort field %q", field)
		}
		seen[field] = struct{}{}

		levels = append(levels, SortLevel{Field: field, Direction: Direction(dir)})
	}

	return levels, nil
}

// Serialize renders a level slice back into the URL-param form. The zero/nil
// slice serializes to the empty string.
func Serialize(levels []SortLevel) string {
	if len(levels) == 0 {
		return ""
	}
	parts := make([]string, len(levels))
	for i, l := range levels {
		parts[i] = fmt.Sprintf("%s:%s", l.Field, l.Direction)
	}
	return strings.Join(parts, ",")
}

File: app/libraries/sortSpec.ts
/**
 * Parses, validates, and serializes the admin lists' multi-level sort specs
 * (e.g. "side:asc,name:asc"). Mirrors pkg/sortspec (Go): keep the grammar in
 * sync. Which field tokens are valid is a per-list concern, so the caller passes
 * the allowed set (PARTY_SORT_FIELDS / GUEST_SORT_FIELDS in the admin options).
 */

export type SortDirection = "asc" | "desc";

export interface SortLevel {
  field: string;
  direction: SortDirection;
}

/** Hard cap matching pkg/sortspec.MaxLevels. */
export const MAX_SORT_LEVELS = 8;

function isSortDirection(s: string): s is SortDirection {
  return s === "asc" || s === "desc";
}

/**
 * Parse a serialized sort spec against the given valid-field set. Returns null
 * for any invalid input (unknown field, bad direction, duplicate, empty pair,
 * whitespace, too many levels); callers treat null as "no sort" and fall back to
 * their default. Mirrors sortspec.Parse in Go (which errors where this returns
 * null).
 */
export function parseSortSpec(
  s: string,
  validFields: ReadonlySet<string>,
): SortLevel[] | null {
  if (!s) return null;
  if (/\s/.test(s)) return null;

  const parts = s.split(",");
  if (parts.length > MAX_SORT_LEVELS) return null;

  const levels: SortLevel[] = [];
  const seen = new Set<string>();

  for (const part of parts) {
    if (!part) return null;
    // Unbounded split + length check so trailing junk ("name:asc:extra")
    // rejects instead of being silently truncated, mirroring Go's SplitN(_, 2)
    // plus direction validation.
    const pieces = part.split(":");
    if (pieces.length !== 2) return null;
    const [field, direction] = pieces;
    if (!field || !direction) return null;
    if (!validFields.has(field)) return null;
    if (!isSortDirection(direction)) return null;
    if (seen.has(field)) return null;
    seen.add(field);
    levels.push({ field, direction });
  }

  return levels;
}

/**
 * Serialize a spec back into the URL-param form. Returns "" for an empty array;
 * callers treat empty output as "no sort" and omit the param.
 */
export function serializeSortSpec(levels: readonly SortLevel[]): string {
  return levels.map((l) => `${l.field}:${l.direction}`).join(",");
}

/** Deep equality for two specs (order matters). */
export function sortSpecsEqual(
  a: readonly SortLevel[] | null | undefined,
  b: readonly SortLevel[] | null | undefined,
): boolean {
  if (!a || !b) return !a && !b;
  if (a.length !== b.length) return false;
  return a.every(
    (l, i) => l.field === b[i].field && l.direction === b[i].direction,
  );
}
